### ISSUE TITLE
Scale in sonar widget not correct

### ISSUE THREAD

**Issue by alcantara09** [rock-core/gui-rock_widget_collection#17]
Comparing the data from the sonar samples and the visualization in the sonar widget, it is possible to notice that the visualization is not correct. I would say that is 10 times smaller.

**Comment by romulogcerqueira** [comment]
@alcantara09 , could you give us more details?

**Comment by alcantara09** [comment]
@romulogcerqueira , 
the error is in the sonar widget. In fact, I think the range is not being applied to the visualization. No matter which range you choose in the widget, the readings from the sonar remains at the same scale. 
![range_bellow_1m](https://cloud.githubusercontent.com/assets/6361620/20800481/c9aaa380-b7cc-11e6-8f8a-339da20370f8.png)
![range_60m](https://cloud.githubusercontent.com/assets/6361620/20800493/d8823918-b7cc-11e6-96c6-bb6c5b14c806.png)

and the script to that produces the error is just running the sonar simulation component and the sonar widget as this one

`require 'rock/bundles'
require 'vizkit'
require 'rock/gazebo'

include Orocos
Bundles.initialize

model_path = File.join(Dir.pwd, 'models','sdf')
world_path = File.join(Dir.pwd, 'scenes','uw_structures', 'uw_structures.world')

## Execute the task ##
Bundles.run 'imaging_sonar_simulation::ScanningSonarTask' => 'sonar_sim'  do

    ## Get the specific task context ##
    sonar_sim = Bundles.get 'sonar_sim'
    #set an array with model paths
    sonar_sim.model_paths = [model_path]
    #path to world file
    sonar_sim.world_file_path = world_path
    ## Configure the tasks ##
    sonar_sim.configure

    sonar_gui = Vizkit.default_loader.SonarWidget
    sonar_sim.sonar_samples.connect_to sonar_gui 

    ## Start the tasks ##
    sonar_sim.start

    #input with at the origin with 0 velocities
    rbs = Types::Base::Samples::RigidBodyState.new
    rbs.position = Eigen::Vector3.new(0,0,0)
    rbs.orientation = Eigen::Quaternion.new(0,0,0,1)

    sim_pose_writer = sonar_sim.sonar_pose_cmd.writer
    timer = Qt::Timer.new
    timer.connect(SIGNAL('timeout()')) do
        sim_pose_writer.write(rbs)
    end
    
    # start the timer with a timeout of 10 ms
    timer.start(10)
    sonar_gui.show
    Vizkit.exec 
end`

**Comment by romulogcerqueira** [comment]
Hi @alcantara09,

I had a look at your script code and apparently it is not a problem in sonar widget and/or sonar simulation. You need to connect the task properties with the sonar widget sliders. Please perform the following steps:

* Insert these lines after `sonar_sim.sonar_samples.connect_to sonar_gui `:
<pre>
  ## Set the signals ##
  sonar_gui.connect(SIGNAL('rangeChanged(int)')) do |value|
    sonar_sim.range = value
  end
  
  sonar_gui.connect(SIGNAL('gainChanged(int)')) do |value|
    sonar_sim.gain = value * 1.0 / 100.0
  end
</pre>

* If the problem is still happenning, open `rock-display` and try to change the properties `range` and `gain` manually.

Please let me know your attempts.

**Comment by doudou** [comment]
> You need to connect the task properties with the sonar widget sliders. Please perform the following steps:

You are not computing the ranges from the sonar data ?

**Comment by romulogcerqueira** [comment]
> You are not computing the ranges from the sonar data ?

Sorry @doudou, I didn't understand your question. `range` is a property of `imaging_sonar_simulation`. If we wish to change its value runtime using script and sonar widget, we need to set the signals of sonar widget gui.

**Comment by doudou** [comment]
The max range is in effect embedded in the sonar data structure (bin_duration * bin_count * speed_of_sound). IMO visualization should match what's there. The max_range control should really only be that, a control.

**Comment by doudou** [comment]
The problem here is that the sonar widget is completely broken when used in e.g. rock-display, which is still the primary use case, and due to vizkit limitations there's no way to automatically plug the range to the sonar component when opening a sonar display.

**Comment by romulogcerqueira** [comment]
> The max range is in effect embedded in the sonar data structure (bin_duration * bin_count * speed_of_sound). 

I agree with you. This makes sense in the most sonar devices (like Tritech DST Micron), however Tritech Gemini and Teledyne Blueview sonars have a range compression which limits the bin_count between 750 -1500 bins, and its value is not proportional to range one (e.g. we can have 1350 bins at 20 m and 900 bins at 25m).

**Comment by romulogcerqueira** [comment]
> IMO visualization should match what's there. The max_range control should really only be that, a control.

In this case, when the range value is changed, the bin_count will also be changed, right? Currently in sonar simulation the bin_count is not dependent of range value. I can fix it now we have a dynamic shader image.

**Comment by doudou** [comment]
Range is controlled by two parameters, the bin_count and the time-per-bin. The latter is usually what's used to control, since one likely uses the max. bin count of the sonar (to get the best resolution)

**Comment by doudou** [comment]
(I'm not even sure that one can change the bin count in practice .... I don't see any benefit in doing so)

**Comment by romulogcerqueira** [comment]
> Range is controlled by two parameters, the bin_count and the time-per-bin. The latter is usually what's used to control, since one likely uses the max. bin count of the sonar (to get the best resolution)
>(I'm not even sure that one can change the bin count in practice .... I don't see any benefit in doing so)

My mistake, sorry! As you said, we already have this relationship in the sonar simulation [here](https://github.com/Brazilian-Institute-of-Robotics/simulation-gpu_sonar_simulation/blob/master/src/Sonar.cpp#L42) to define `bin_duration`. It is not needed to change the `bin_count` when the `range` is changed.

**Comment by romulogcerqueira** [comment]
@doudou , I'm in vacations in this month, however I will try to keep tracking here as soon as possible.
